File: drills/elixir/solutions/nc56_rotate_image.exs
defmodule Solution do
  # A quarter turn is a reflection through the main diagonal followed by a
  # reflection through the vertical centre line. Two easy operations instead of
  # one four-way element cycle, and neither needs index arithmetic.
  def rotate([]), do: []

  def rotate(matrix) do
    matrix
    |> Enum.zip_with(& &1)
    |> Enum.map(&Enum.reverse/1)
  end
end

File: drills/ts/solutions/nc56_rotate_image.ts
export function rotate(matrix: number[][]): number[][] {
  // A quarter turn is a reflection through the main diagonal followed by a
  // reflection through the vertical centre line. Two easy operations instead of
  // one four-way element cycle, and neither needs index arithmetic.
  if (matrix.length === 0) return [];
  const transposed = matrix[0].map((_, c) => matrix.map((row) => row[c]));
  return transposed.map((row) => row.reverse());
}
